### PR TITLE
Fix lightning and pytorch estimator loading from checkpoint

### LINKED ISSUE #3268: Unable to load most recent checkpoint for Pytorch and Pytorch lightning Estimator
**Environment:**
1. Framework: PyTorch
2. Framework version: 1.8.1
3. Horovod version: 0.23.0
4. MPI version:
5. CUDA version:
6. NCCL version:
7. Python version: 3.8
8. Spark / PySpark version: 3.1.2
9. Ray version:
10. OS and version:
11. GCC version:
12. CMake version:

**Bug report:**
In case of pytorch lightning estimator, the _read_checkpoint() API does not return the latest checkpoint stored in the run path. 
Reason: Pytorch lightning estimator calls store.get_checkpoints() which looks for a folder named 'checkpoint' in run path while there is no folder named checkpoint, instead there is a temp folder generated via tempfile.TemporaryDirectory()

In case of pytorch estimator, the checkpoint stored in run path is not overwritten if multiple iterations are done using the same run path, which leads to _load_checkpoint() API returning the stale checkpoint.

### PR DESCRIPTION
**Pull request by kamalsharma2**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes #3268. Creating named temporary output directories which are later copied to run path, then fetching the checkpoints from this named directory in run path to ensure latest checkpoints are fetched in multiple iterations.

### CODE REVIEW

File: horovod/spark/lightning/remote.py [hunk 1]
```diff
@@ -262,7 +262,7 @@ def on_epoch_end(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -
             if hvd.rank() == 0:
                 if remote_store.saving_runs and trainer.profiler:
                     # One more file sync to push profiler result.
-                    remote_store.sync(logs_path)
+                    remote_store.sync(run_output_dir)
```
Comment (chongxiaoc) [inline horovod/spark/lightning/remote.py:265]: Checkpoint is synced at every epoch.
No need to sync checkpoint files again here.
This is only for uploading profiler results, which is available at the end of training.

File: horovod/spark/lightning/remote.py [hunk 2]
```diff
@@ -111,7 +111,7 @@ def train(serialized_model):
         _checkpoint_callback = None
         require_checkpoint = False
 
-        with remote_store.get_local_output_dir() as run_output_dir:
+        with remote_store.get_local_run_dir() as run_output_dir:
```
Comment (chongxiaoc) [inline horovod/spark/lightning/remote.py:114]: Why the existing `get_local_output_dir()` is not working here? 
It is an existing local tmp path.

**Comment by chongxiaoc** [comment]
I left two comments open to discussion.
@kamalsharma2 
Thanks for fixing it.

File: horovod/spark/lightning/remote.py (hunk 2 above)
Comment (kamalsharma2) [reply horovod/spark/lightning/remote.py:114]: `get_local_output_dir()` creates a random temporary directory which is then copied to run path, but when loading checkpoints it looks for a specific directory named 'checkpoint' which is not created, hence this change.

File: horovod/spark/lightning/remote.py (hunk 1 above)
Comment (kamalsharma2) [reply horovod/spark/lightning/remote.py:265]: Got it, will revert this change.

File: horovod/spark/lightning/remote.py (hunk 2 above)
Comment (chongxiaoc) [reply horovod/spark/lightning/remote.py:114]: got it. thx!

**Review by chongxiaoc — APPROVED** [review]
LGTM. Thx for fixing.

**Comment by chongxiaoc** [comment]
@irasit Do you have any comments?

**Review by irasit — APPROVED** [review]

**Comment by irasit** [comment]
LGTM

File: horovod/spark/common/store.py [hunk 3]
```diff
@@ -281,6 +295,17 @@ def local_run_path():
                 yield tmpdir
         return local_run_path
 
+    def get_local_run_dir_fn(self, run_id):
+        @contextlib.contextmanager
+        def local_run_path():
+            dirpath = os.path.join(tempfile._get_default_tempdir(),self.get_checkpoint_dir)
```
Comment (EnricoMi) [inline horovod/spark/common/store.py:301]: this does not call into `get_checkpoint_dir` but uses the method as the argument, causing:

```
Tue Nov 23 11:59:00 2021[1,1]<stderr>:  File "/usr/local/lib/python3.8/dist-packages/horovod/spark/common/store.py", line 301, in local_run_path
Tue Nov 23 11:59:00 2021[1,1]<stderr>:    dirpath = os.path.join(tempfile._get_default_tempdir(),self.get_checkpoint_dir)
Tue Nov 23 11:59:00 2021[1,1]<stderr>:  File "/usr/lib/python3.8/posixpath.py", line 90, in join
Tue Nov 23 11:59:00 2021[1,1]<stderr>:    genericpath._check_arg_types('join', a, *p)
Tue Nov 23 11:59:00 2021[1,1]<stderr>:  File "/usr/lib/python3.8/genericpath.py", line 152, in _check_arg_types
Tue Nov 23 11:59:00 2021[1,1]<stderr>:    raise TypeError('%s() argument must be str or bytes, not %r' %
Tue Nov 23 11:59:00 2021[1,1]<stderr>:TypeError: join() argument must be str or bytes, not 'method'
```
https://github.com/horovod/horovod/runs/4298078548?check_suite_focus=true#step:163:91

**Comment by EnricoMi** [comment]
@kamalsharma2 can you please provide a PR fixing https://github.com/horovod/horovod/pull/3279/files#r755062520?

**Comment by kamalsharma2** [comment]
@EnricoMi raised a PR for the typo

**Comment by kamalsharma2** [comment]
@EnricoMi, raised a PR for the typo, ignore the closed one

**Comment by EnricoMi** [comment]
thanks @kamalsharma2, see you there

**Pull request merged**